Doc: Add release note for v0.5 release. (#2287)
projectacrn/acrn-hypervisor: doc/release_notes_0.5.rst

####################

We are pleased to announce the release of Project ACRN version 0.5.

ACRN is a flexible, lightweight reference hypervisor, built with
real-time and safety-criticality in mind, optimized to streamline
embedded development through an open source platform. Check out the
:ref:`introduction` for more information.

All project ACRN source code is maintained in the
https://github.com/projectacrn/acrn-hypervisor repository and includes
folders for the ACRN hypervisor, the ACRN device model, and documentation.
You can either download this source code as a zip or tar.gz file (see
the `ACRN v0.5 GitHub release page
<https://github.com/projectacrn/acrn-hypervisor/releases/tag/v0.5>`_ or
use Git clone and checkout commands:

.. code-block:: bash

   git clone https://github.com/projectacrn/acrn-hypervisor
   cd acrn-hypervisor
   git checkout v0.5

The project's online technical documentation is also tagged to correspond
with a specific release: generated v0.5 documents can be found at
https://projectacrn.github.io/0.5/.  Documentation for the latest
(master) branch is found at https://projectacrn.github.io/latest/.

Version 0.5 new features
************************

> - :acrn-issue:`xxxx` - Support SBL firmware as boot loader on Apollo Lake UP2.

Reviewer: needs an actual issue number (not xxxx)

> - :acrn-issue:`1832` - Add OVMF booting support for booting as an alterative to vSBL.

Reviewer: ```suggestion
- :acrn-issue:`1832` - Add OVMF booting support for booting as an alternative to vSBL.
```

- :acrn-issue:`2170` -  For UEFI based hardware platforms, one ClearLinux E2E build binary can be used for all platform's installation 
- :acrn-issue:`2036` - Update and complete `acrn-dm` parameters description in the user guide and HLD

> - :acrn-issue:`1821` - Optimize IO request path  

Reviewer: Can you sort this by issue number please.

   **Workaround:** None.

:acrn-issue:`2267` - [APLUP2][LaaG]LaaG can't detect 4k monitor 
   After launching UOS on APL UP2 , 4k monitor cannot be detected.

   **Impact:** UOS hasn't display with 4k monitor.

   **Workaround:** None.

:acrn-issue:`2276` - OVMF failed to launch UOS on UP2.
   UP2 failed to launch UOS using OVMF as virtual bootloader with acrn-dm. 

   **Impact:** UOS cannot boot up using OVMF

   **Workaround:** Use VSBL as virtual bootloader

:acrn-issue:`2277` - [APLNUC]Launch UOS with 5G memory will hang 2 minutes
   If launching UOS with 5G memory, there will be about 2 minutes hang.

   **Impact:** Low UOS boot time performance.

   **Workaround:** None.

:acrn-issue:`2278` - [KBLNUC] Cx/Px is not supported on KBLNUC
   C states/P states is not supported on KBL NUC.

   **Impact:** Power Management states related operations cannot be using in SOS/UOS on KBLNUC

   **Workaround:** None 

:acrn-issue:`2279` - [APLNUC]After exiting UOS with mediator Usb_KeyBoard and Mouse, SOS cannot use the 
   Usb_KeyBoard and Mouse
   After exiting UOS with mediator Usb_KeyBoard and Mouse, SOS cannot use the Usb_KeyBoard and Mouse.
   Reproduce Steps as below:

>    1: Insert USB keyboard and mouse in standard A port（USB3.0 port）

Reviewer: To be consistent, use a numbered list as above:
```
   1) Insert USB keyboard and mouse in standard A port （USB3.0 port）
   2) Boot UOS by sharing the USB keyboard and mouse in cmd line::

         -s n,xhci,1-1:1-2:1-3:1-4:2-1:2-2:2-3:2-4 \

   3) UOS access USB keyboard and mouse.
   4) Exit UOS.
   5) SOS access USB keyboard and mouse.
```

   2: Boot UOS by sharing the USB keyboard and mouse in cmd line 
   -s n,xhci,1-1:1-2:1-3:1-4:2-1:2-2:2-3:2-4 \

   3: UOS access USB keyboard and mouse.

   4: Exit UOS.

   5: SOS access USB keyboard and mouse. 

   **Impact:** SOS cannot use USB keyboard and mouse in such case.

   **Workaround:** Unplug and plug-in the USB keyboard and mouse after exiting UOS.

.. comment
   Use the syntax:

   :acrn-issue:`663` - Short issue description
     Longer description that helps explain the problem from the user's
     point of view (not internal reasons).  **Impact:** What's the
     consequences of the issue, and how it can affect the user or system.
     **Workaround:** Describe a workaround if one exists (or refer them to the
     :acrn-issue:`663`` if described well there. If no workaround, say
     "none".

Change Log
**********

These commits have been added to the acrn-hypervisor repo since the v0.4

> release in Jan 2019 (click on the CommitID link to see details):

Reviewer: ```suggestion
release in Dec 2018 (click on the CommitID link to see details):
```

.. comment

   This list is obtained from the command:
   git log --pretty=format:'- :acrn-commit:`%h` %s' --after="2018-03-01"

- :acrn-commit:`b7fda274` config: fix no serial output with SBL on UP2
- :acrn-commit:`ddf1c923` hv: fix violations in md.c md.h and md_internal.h for crypto lib
- :acrn-commit:`c230a1a6` hv: fix violations in sha256.c for crypto lib
- :acrn-commit:`488e7b2a` hv: fix violations in hkdf.c and crypto_api.c for crypto lib

> - :acrn-commit:`08843973` HV: cycomatic complexity 20 in vlapic_icrlo_write_handler

Reviewer: ```suggestion
- :acrn-commit:`08843973` HV: cyclomatic complexity 20 in vlapic_icrlo_write_handler
```

- :acrn-commit:`eaa0e307` HV: remove multiple exit/return in routines in the file of vlapic.c
- :acrn-commit:`8e00180c` HV: Remove goto statement in guest.c

> - :acrn-commit:`2e01b4c8` HV: trival changes to meet MISRA

Reviewer: ```suggestion
- :acrn-commit:`2e01b4c8` HV: trivial changes to meet MISRA-C
```

> - :acrn-commit:`971eb84c` HV: add const qualifier for functions' argments in vlapic.c

Reviewer: ```suggestion
- :acrn-commit:`971eb84c` HV: add const qualifier for functions' arguments in vlapic.c
```

- :acrn-commit:`067273af` hv: assign: fix remaining MISRA-C violations
- :acrn-commit:`1dfd05cd` hv: fix mis-usage of "PAGE_SHIFT"
- :acrn-commit:`5c6fe01c` hv:Change pcpu_active_bitmap to static
- :acrn-commit:`682824de` hv:Change phys_cpu_num to static
- :acrn-commit:`59e2de48` dm: acpi: add PSDS table in ACPI table
- :acrn-commit:`90fd5d58` script: fix launch_uos script issue due to unseen character
- :acrn-commit:`96800093` doc: update footer copyright year
- :acrn-commit:`9c27ed10` profiling: fix the profiling tool crash by page faults
- :acrn-commit:`a177d75e` doc: initial draft of ACRN coding guidelines
- :acrn-commit:`d89ce8ae` hv: schedule: fix "Procedure has more than one exit point"
- :acrn-commit:`952943c3` hv: decouple IO completion polling from idle thread
- :acrn-commit:`a0154223` hv: clear NEED_RESCHEDULE flag in schedule
- :acrn-commit:`e8ac9767` hv: use asm_pause() to replace inline ASM to satisfy MISRAC
- :acrn-commit:`329ea42d` dm: fix the memory leak in virtio mei
- :acrn-commit:`8f57c61d` dm: Add teardown callback for mevent in uart_core
- :acrn-commit:`72d1fa50` dm: refine the uart_core
- :acrn-commit:`21aa1907` hv: vcpuid: cpuid leaf 07h has subleaf

> - :acrn-commit:`2d3f510d` hv: tursty_hypercall: fix "Procedure has more than one exit point"

Reviewer: ```suggestion
- :acrn-commit:`2d3f510d` hv: trusty_hypercall: fix "Procedure has more than one exit point"
```

- :acrn-commit:`ddd07b95` hv: cpu_state_tbl: fix multiple exits
- :acrn-commit:`eb77e25f` hv: ept: fix MISRA-C violations
- :acrn-commit:`5253ac7a` dm: virtio: refine header file
- :acrn-commit:`738f2536` hv: coding style: refine cpu related function to one exit
- :acrn-commit:`9672538c` init: move init_scheduler into cpu.c
- :acrn-commit:`ff0703dd` scheduler: make scheduling based on struct sched_object
- :acrn-commit:`8aae0dff` scheduler: refine make_reschedule_request
- :acrn-commit:`6d673648` scheduler: refine runqueue related functions
- :acrn-commit:`93e588bc` hv: fix e820.c violations
- :acrn-commit:`60f78e1e` hv:vtd: fix MISRA-C violations on procedure has more than one exit point
- :acrn-commit:`a98a1a69` hv:vtd: fix MISRA-C violations on pointer not checked for null before use
- :acrn-commit:`725e1921` hv:vtd: fix MISRA-C violations on comment possibly contains code
- :acrn-commit:`897ffa27` hv:vtd: fix MISRA-C violations on logical conjunctions need brackets
- :acrn-commit:`80b392a8` hv:vtd: fix MISRA-C violations on pointer param should be declared pointer to const
- :acrn-commit:`5282fa89` hv:vtd: fix MISRA-C violations on scope of variable could be reduced

> - :acrn-commit:`bec21d14` Patch for modularising ioapic.[c/h] and related files.

Reviewer: ```suggestion
- :acrn-commit:`bec21d14` Patch for modularizing ioapic.[c/h] and related files.
```

- :acrn-commit:`af9b7476` doc: fix formatting in NUC GSG
- :acrn-commit:`61f03dae` DOC: change PCI uart description from mmio to bdf
- :acrn-commit:`50f5b0f6` hv: vmexit: fix MISRA-C violations related to multiple exits
- :acrn-commit:`0a713e6f` hv: coding style: refine set_vcpuid_entries to one exit
- :acrn-commit:`a56abee9` hv: coding style: refine find_vcpuid_entry
- :acrn-commit:`58d2a418` HV: fix pm code for multi-exits & unsigned const
- :acrn-commit:`97a73951` dm: pass mac seed not to use vm name on UP2
- :acrn-commit:`1c99a975` hv: coding style: refine trusty
- :acrn-commit:`1dca17cd` hv: coding style: refine initialize_trusty to one exit
- :acrn-commit:`8a55f038` hv: coding style: refine hcall_initialize_trusty to one exit
- :acrn-commit:`1d1d2434` DM USB: xHCI: change log level of some logs for S3 online debugging
- :acrn-commit:`5f0c093e` hv: coding style: remove no real declaration for external variable

> - :acrn-commit:`1e3358fd` Debug: Add one hypercall to quary hardware info

Reviewer: ```suggestion
- :acrn-commit:`1e3358fd` Debug: Add one hypercall to query hardware info
```

- :acrn-commit:`e22b35e3` HV/DM: Unify the usage of aligned for structure definition with alignment
- :acrn-commit:`71a80d2d` hv: assign: change ptirq vpin source type from enum to macro
- :acrn-commit:`d5865632` hv: assign: remove added ptirq entries if fails to add all
- :acrn-commit:`d48dc387` hv: assign: fix MISRA-C violations on multiple exits
- :acrn-commit:`e8b3e44f` hv: assign: fix MISRA-C violations on potential null pointer deference
- :acrn-commit:`e19dcf57` hv: assign: fix MISRA-C violations on implicit type conversion
- :acrn-commit:`714814f9` hv: move `atoi` and `strtol_dec` to debug directory
- :acrn-commit:`32d6aa97` hv: string: fix MISRA-C violations related to style
- :acrn-commit:`2c6c383e` hv: string: fix MISRA-C violations related to break
- :acrn-commit:`b319e654` HV: fix bug adapt uart mmio to bdf for HV cmdline
- :acrn-commit:`23c2166a` HV: change serial PCI cfg to bus:dev.func format
- :acrn-commit:`1caf58f2` hv:clean io_request.c misra violations
- :acrn-commit:`530388db` hv: irq: fix MISRA-C violations in irq.c and idt.h
- :acrn-commit:`08cf8f64` hv: lapic: fix MISRA-C violation of potential numeric overflow

> - :acrn-commit:`83ebd432` hv: ptdev: fix MISRAC violations

Reviewer: ```suggestion
- :acrn-commit:`83ebd432` hv: ptdev: fix MISRA-C violations
```

- :acrn-commit:`eff94591` HV: x86: fix "Procedure has more than one exit point"
- :acrn-commit:`e283e774` hv: vmcs: fix MISRA-C violations related to multiple exits
- :acrn-commit:`4618a6b1` hv: vmcs: fix MISRA-C violations related to pointer
- :acrn-commit:`8e58a686` hv: vmcs: fix MISRA-C violations related to variable scope
- :acrn-commit:`9a051e7a` hv: vmcs: fix MISRA-C violations related to style
- :acrn-commit:`7d8cd911` security: remove gcc flags Wformat Wformat-security in HV
- :acrn-commit:`d133f95d` hv: fix MISRA-C violations "Pointer param should be declared pointer to const."
- :acrn-commit:`f81fb21a` HV: modularization to refine pm related code.
- :acrn-commit:`03262a96` hv: refine coding style for ucode.c
- :acrn-commit:`927c5172` hv: vpci: fix MISRA-C violations related to variable declarations
- :acrn-commit:`4c28e98d` hv: refine a few functions to only one exit point
- :acrn-commit:`64a46300` hv:refine prepare_vm0 api
- :acrn-commit:`b5e0efca` hv: coding style: refine memory.c
- :acrn-commit:`5b467269` hv: lib: remove memchr
- :acrn-commit:`97132acc` Make ibrs_type as internal variable

> - :acrn-commit:`55cce7e4` Fix Misra C violation in cpu_caps.c and security.c

Reviewer: ```suggestion
- :acrn-commit:`55cce7e4` Fix MISRA-C violation in cpu_caps.c and security.c
```

- :acrn-commit:`689c1c28` function name change in init.c
- :acrn-commit:`5968da46` move security related funcs into security.c
- :acrn-commit:`0ad6da99` make detect_cpu_cap as internal function
- :acrn-commit:`e22217fd` refine apicv capability check
- :acrn-commit:`7c8b7671` refine in cpu_caps.c
- :acrn-commit:`63773db4` change get_monitor_cap to has_monitor_cap

> - :acrn-commit:`6830619d` modulization: combine vmx_caps into cpu_caps

Reviewer: ```suggestion
- :acrn-commit:`6830619d` modularization: combine vmx_caps into cpu_caps
```

> - :acrn-commit:`746fbe14` modulization: move functions related with cpu caps into cpu_caps.c

Reviewer: ```suggestion
- :acrn-commit:`746fbe14` modularization: move functions related with cpu caps into cpu_caps.c
```

- :acrn-commit:`b8ffac8b` hv:fix possible buffer overflow in 'ptirq_get_intr_data()'
- :acrn-commit:`6aa42272` fix "Procedure has more than one exit point."
- :acrn-commit:`65a7be8f` hv:refine alloc_vm_id api
- :acrn-commit:`235ad0ff` hv: refine memcpy_s
- :acrn-commit:`f9897c6f` hv: refine memset
- :acrn-commit:`78e9a84f` hv: add fast string enhanced rep movsb/stosb check on initial
- :acrn-commit:`3515ca1e` hv: vpci: fix "Procedure has more than one exit point"
- :acrn-commit:`c547e9cf` hv: enable/disable snoop control bit per vm
- :acrn-commit:`20280341` hv: MISRA-C fix "identifier reuse" in vpci code
- :acrn-commit:`2ddd24e0` dm: storage: support discard command
- :acrn-commit:`f71370ad` dm: storage: rename delete to discard

> - :acrn-commit:`36863a0b` modulization: vmx on/off should not use vcpu param

Reviewer: ```suggestion
- :acrn-commit:`36863a0b` modularization: vmx on/off should not use vcpu param
```

- :acrn-commit:`bed82dd3` cleanup vmcs source and header files

> - :acrn-commit:`731c4836` modulization: separate vmx.c into two parts

Reviewer: ```suggestion
- :acrn-commit:`731c4836` modularization: separate vmx.c into two parts
```

- :acrn-commit:`0d5c65f1` hv: enforce data size on all out exits
- :acrn-commit:`5ab68eb9` dm: hw: Replace sscanf with permitted string API
- :acrn-commit:`63b814e7` dm: hw: Replace strlen with strnlen
- :acrn-commit:`eab7cd47` dm: hw: Replace sprintf with snprintf

> - :acrn-commit:`69dc9392` hv: drop the temperory stack for AP startup

Reviewer: ```suggestion
- :acrn-commit:`69dc9392` hv: drop the temporary stack for AP startup
```

> - :acrn-commit:`74849cd9` modulization:move out efi dir from hypervisor

Reviewer: ```suggestion
- :acrn-commit:`74849cd9` modularization:move out efi dir from hypervisor
```

- :acrn-commit:`4f36244f` dm: virtio-console: apply new mevent API to avoid race issue
- :acrn-commit:`baf8f8bd` dm: virtio-input: apply new mevent API to avoid race issue
- :acrn-commit:`c2df4a85` DM USB: xHCI: no wait logic implementation for S3
- :acrn-commit:`82659831` DM USB: xHCI: refine emulation for command XHCI_CMD_RS
- :acrn-commit:`e5c98e6d` DM USB: add usb_dev_path_cmp function for convenience
- :acrn-commit:`6c1ca137` DM USB: xHCI: remove the waiting 5 seconds wa for s3
- :acrn-commit:`4fc5dcfc` hv: enable SMAP in hypervisor
- :acrn-commit:`57dfc7de` hv: refine IOREQ state operation functions in hypervisor
- :acrn-commit:`c89d6e65` modularization: clean up namings in vMTRR module
- :acrn-commit:`6bbd0129` modularization: move vMTRR code to guest directory
- :acrn-commit:`e066774a` hv: refine strnlen_s/strstr_s to only one exit point
- :acrn-commit:`e114ea7e` hv: timer: fix procedure has more than one exit point
- :acrn-commit:`4131d46f` hv: remove goto in ept_violation_vmexit_handler
- :acrn-commit:`a958fea7` hv: emulate IA32_TSC_ADJUST MSR
- :acrn-commit:`6b998580` Fix KW issues for tpm_emulator

> - :acrn-commit:`2d469a5e` modularization: hypervisor initilization component

Reviewer: ```suggestion
- :acrn-commit:`2d469a5e` modularization: hypervisor initialization component
```

- :acrn-commit:`2d9e478c` Doc: delete the step of downloading UOS's kernel
- :acrn-commit:`c3a4a5d4` Doc: add "$" for code
- :acrn-commit:`d56e2c29` Doc: update the steps
- :acrn-commit:`2be939f3` Doc: add "Deploy the UOS kernel modules for AGL"
- :acrn-commit:`73161f91` Update using_agl_as_uos.rst
- :acrn-commit:`c51394c3` doc: update the doc of AGL as UOS
- :acrn-commit:`e5748795` doc: update the doc of using AGL as UOS
- :acrn-commit:`fbaecde6` DM USB: xHCI: Fix banned API issue.
- :acrn-commit:`e835f5f5` dm: enforce data size when accessing PCI BARs
- :acrn-commit:`f5a66e8e` doc: update OVMF usage for acrn-dm
- :acrn-commit:`d8c4e7d3` dm: add option to boot OVMF from acrn-dm
- :acrn-commit:`9e97fd06` dm: add BIOS/ROM image loading support at High BIOS region
- :acrn-commit:`653a5795` dm: query and save image size during initial checking
- :acrn-commit:`a80f08eb` dm: add launch_uos.args sample file for AaaG
- :acrn-commit:`04fef4f3` tools: acrn-manager: change path of vm conf files
- :acrn-commit:`2f30dcdb` hv: refine strncpy_s to only one exit point
- :acrn-commit:`b8ca17c6` hv: remove strcpy_s
- :acrn-commit:`29c8494f` hv: replace strcpy_s with strncpy_s

> - :acrn-commit:`07427b4c` modulization: move virtual cpuid stuff into guest dir

Reviewer: ```suggestion
- :acrn-commit:`07427b4c` modularization: move virtual cpuid stuff into guest dir
```

ACRN is a flexible, lightweight reference hypervisor, built with
real-time and safety-criticality in mind, optimized to streamline
embedded development through an open source platform. Check out the
:ref:`introduction` for more information.

All project ACRN source code is maintained in the
https://github.com/projectacrn/acrn-hypervisor repository and includes
folders for the ACRN hypervisor, the ACRN device model, and documentation.
You can either download this source code as a zip or tar.gz file (see
the `ACRN v0.5 GitHub release page
<https://github.com/projectacrn/acrn-hypervisor/releases/tag/v0.5>`_ or
use Git clone and checkout commands:

.. code-block:: bash

   git clone https://github.com/projectacrn/acrn-hypervisor
   cd acrn-hypervisor
   git checkout v0.5

The project's online technical documentation is also tagged to correspond
with a specific release: generated v0.5 documents can be found at
https://projectacrn.github.io/0.5/.  Documentation for the latest
(master) branch is found at https://projectacrn.github.io/latest/.

Version 0.5 new features
************************

**OVMF support initial patches merged in ACRN**: 

> In order to boot Guest Windows on ACRN, the virtual machines firmware is 

Reviewer: Change to:
```
We recommend using Opensource Virtual Machine Firmware (OVMF) when
booting Windows as a guest OS on ACRN.
```

suggested to use OVMF, which is Opensource Virtual Machine Firmware. 
Initial patches to support OVMF have been merged in ACRN hypervisor. 
More patches for ACRN and patches upstreaming to OVMF work will be continuing.

**UP2 board serial port support**: 

> Enabled UP2 board’s serial port debugging during SOS and UOS boot.

Reviewer: Change to:
```
This release enables serial port debugging on UP2 boards during SOS and UOS boot.
```

> One E2E binary to support all UEFI platform: ACRN can support 

Reviewer: This looks like it may be new item.  If so change to:
```
**One E2E binary to support all UEFI platform**: 
ACRN can support both ApolloLake (APL) and KabbyLake (KBL) NUCs.
Instead of having separate builds, this release offers community
developers a single end-to-end reference build that supports both
UEFI hardware platforms, configured with a new boot parameter.
See :ref:`getting_started` for more information.
```

ApolloLake NUC and Kabylake NUC. It’ll be easier to use if 
we can have an all-in-one E2E build to support all UEFI hardware platforms. 
From this release, community developers can use one end-to-end reference 
build to run on both APL NUC and KBL NUC, only need a boot parameter 
change in `Getting Startup Guide <https://projectacrn.github.io/latest/getting-started/index.html>`_

> **APL UP2 board with SBL firmware**: ACRN can support APL UP2 board with 

Reviewer: Change to:
```
**APL UP2 board with SBL firmware**:  With this 0.5 release, ACRN
now supports APL UP2 board with slim Bootloader (SBL) firmware.
```

Slim Bootloader firmware now from this release. Slim Bootloader is a modern, 
flexible, light-weither, open source reference boot loader with key benefits 

> like, fast, small, customizable, secure. An end-to-end reference build with 

Reviewer: ```
such as being fast, small, customizable, and secure.
```

> ACRN hypervisor, Clear Linux as SOS, Clear Linux as UOS has been tried and 

Reviewer: ```suggestion
ACRN hypervisor, Clear Linux as SOS, and Clear Linux as UOS has been 
```

> verified on UP2/SBL board. A step-by-step Getting Startup Guide (GSG) has 

Reviewer: ```
See the :ref:`using-sbl-up2` documentation for step-by-step instructions.
```

been posted in ACRN documentation website.

**Document update**: Several documents have been added or updated in this release. 
“GPU passthrough guidance on Skylake NUC”, “acrn-dm parameters introduction”. 

> One page ACRN project brochure has been uplifted and able to download.

Reviewer: Change to:
```
**Document updates**: Several new documents have been added in this release, including:

* :ref:`skl-nuc-gpu-passthrough`
* :ref:`acrn-dm_parameters`
```

Where is this ACRN project brochure uploaded?

Author: ACRN-dm parameter updated in Issue 2036,
